images/providers.go

```diff
@@ -0,0 +1,100 @@
+package images
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/models"
+	"github.com/prometheus/alertmanager/types"
+)
+
+// ErrImageUploadNotSupported is returned when image uploading is not supported.
+var ErrImageUploadNotSupported = errors.New("image upload is not supported")
+
+type UnavailableProvider struct{}
+
+var _ Provider = (*UnavailableProvider)(nil)
+
+func (u *UnavailableProvider) GetImage(_ context.Context, _ types.Alert) (*Image, error) {
+	return nil, ErrImagesUnavailable
+}
+
+// URLProvider is a provider that stores a direct reference to an image's public URL in an alert's annotations.
+// The URL is not validated against a database record, so retrieving raw image data is blocked in an attempt
+// to prevent malicious access to untrusted URLs.
+type URLProvider struct{}
+
+var _ Provider = (*URLProvider)(nil)
+
+// GetImage returns the image associated with a given alert.
+// The URL should be treated as untrusted and notifiers should pass the URL directly without attempting to download
+// the image data.
+func (u *URLProvider) GetImage(_ context.Context, alert types.Alert) (*Image, error) {
+	url := GetImageURL(alert)
+	if url == "" {
+		return nil, nil
+	}
+
+	return &Image{
+		URL: url,
+		RawData: func(_ context.Context) (ImageContent, error) {
+			// Raw images are not available for alerts provided directly by annotations as the image data is non-local.
+			// While it might be possible to download the image data, it's generally not safe to do so as the URL is
+			// not guaranteed to be trusted.
+			return ImageContent{}, ErrImageUploadNotSupported
+		},
+	}, nil
+}
+
+type TokenStore interface {
+	GetImage(ctx context.Context, token string) (*Image, error)
+}
+
+// TokenProvider is an implementation of the Image Provider interface that retrieves images from a store
+// using tokens.
+// Image data should be considered trusted by the fact that the stored image url and content and not user-modifiable.
```

Comment:
`ImageProvider` without spaces, avoiding the line break before "using tokens", uppercasing "URL", "and" -> "are".
```suggestion
// TokenProvider is an implementation of the ImageProvider interface that retrieves images from a store using tokens.
// Image data should be considered trusted by the fact that the stored image URL and content and not user-modifiable.
```

Or more concisely:
```suggestion
// TokenProvider implements the ImageProvider interface, retrieving images from a store using tokens.
// Image data should be considered trusted as the stored image URL and content are not user-modifiable.
```